File: internal/model/advance_payment.go
package model

import (
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

type AdvancePaymentStatus string

const (
	AdvancePaymentStatusDraft             AdvancePaymentStatus = "draft"
	AdvancePaymentStatusConfirmed         AdvancePaymentStatus = "confirmed"
	AdvancePaymentStatusPartiallyAllocated AdvancePaymentStatus = "partially_allocated"
	AdvancePaymentStatusFullyAllocated    AdvancePaymentStatus = "fully_allocated"
	AdvancePaymentStatusReversed          AdvancePaymentStatus = "reversed"
)

type AdvancePayment struct {
	ID               uuid.UUID       `db:"id"`
	TenantID         uuid.UUID       `db:"tenant_id"`
	CompanyID        uuid.UUID       `db:"company_id"`
	SupplierID       uuid.UUID       `db:"supplier_id"`
	AdvanceNo        string          `db:"advance_no"`
	Amount           decimal.Decimal `db:"amount"`
	AllocatedAmount  decimal.Decimal `db:"allocated_amount"`
	OutstandingAmount decimal.Decimal `db:"outstanding_amount"`
	PaidDate         time.Time       `db:"paid_date"`
	DueDate          *time.Time      `db:"due_date"`
	Status           string          `db:"status"`
	SourceType       string          `db:"source_type"`
	BankAccountID    *uuid.UUID      `db:"bank_account_id"`
	ReferenceNo      *string         `db:"reference_no"`
	Remark           *string         `db:"remark"`
	VoucherID        *uuid.UUID      `db:"voucher_id"`
	VoucherNo        *string         `db:"voucher_no"`
	CreatedBy        *uuid.UUID      `db:"created_by"`
	CreatedAt        time.Time       `db:"created_at"`
	ConfirmedBy      *uuid.UUID      `db:"confirmed_by"`
	ConfirmedAt      *time.Time      `db:"confirmed_at"`
	ReversedBy       *uuid.UUID      `db:"reversed_by"`
	ReversedAt       *time.Time      `db:"reversed_at"`
}

File: frontend/src/api/modules/advance_payment.ts
import request from '@/api/request'
import type { ApiResponse } from '@/types/api'

export interface AdvancePayment {
  id: string
  advance_no: string
  supplier_id: string
  amount: string
  allocated_amount: string
  outstanding_amount: string
  paid_date: string
  due_date?: string
  status: string
  source_type: string
  bank_account_id?: string
  reference_no?: string
  remark?: string
  voucher_no?: string
  created_at: string
  confirmed_at?: string
}

export interface CreateAdvancePaymentPayload {
  company_id: string
  supplier_id: string
  amount: string
  paid_date: string
  due_date?: string
  bank_account_id?: string
  reference_no?: string
  remark?: string
}

export function fetchAdvancePayments(params: { status?: string } = {}): Promise<ApiResponse<{ list: AdvancePayment[]; total: number }>> {
  return request.get('/advance-payments', { params })
}

export function fetchAdvancePaymentById(id: string): Promise<ApiResponse<AdvancePayment>> {
  return request.get(`/advance-payments/${id}`)
}

export function fetchOutstandingAdvancePayments(supplierId: string): Promise<ApiResponse<{ list: AdvancePayment[]; total: number }>> {
  return request.get('/advance-payments/outstanding', { params: { supplier_id: supplierId } })
}

export function createAdvancePayment(payload: CreateAdvancePaymentPayload): Promise<ApiResponse<AdvancePayment>> {
  return request.post('/advance-payments', payload)
}

export function confirmAdvancePayment(id: string): Promise<ApiResponse<AdvancePayment>> {
  return request.post(`/advance-payments/${id}/confirm`)
}
